services/src/main/java/org/exoplatform/officeonline/WOPIService.java

```diff
@@ -361,7 +449,79 @@ public void putFile(EditorConfig config, String lockId, InputStream data) throws
       throw new OfficeOnlineException("Cannot perform putFile operation. FileId: " + config.getFileId() + ", workspace: "
           + config.getWorkspace());
     }
+  }
+
+  /**
+   * Gets the versioning user.
+   *
+   * @param frozen the frozen
+   * @return the versioning user
+   * @throws RepositoryException the repository exception
+   */
+  protected String getVersioningUser(Node frozen) throws RepositoryException {
+    if (frozen.hasProperty(WOPIService.MSOFFICE_VERSION_OWNER)) {
+      return frozen.getProperty(WOPIService.MSOFFICE_VERSION_OWNER).getString();
+    }
+    return null;
+  }
+
+  /**
+   * Gets the frozen.
+   *
+   * @param node the node
+   * @return the frozen
+   * @throws RepositoryException the repository exception
+   */
+  protected Node getFrozen(Node node) throws RepositoryException {
+    if (node.getBaseVersion().hasNode(WOPIService.JCR_FROZEN_NODE)) {
+      return node.getBaseVersion().getNode(WOPIService.JCR_FROZEN_NODE);
+    }
+    return node;
+  }
+
+  /**
+   * Checks if is editor version.
+   *
+   * @param frozen the frozen
+   * @return the boolean
+   * @throws RepositoryException the repository exception
+   */
+  protected Boolean isEditorVersion(Node frozen) throws RepositoryException {
+    if (frozen.hasProperty(WOPIService.MSOFFICE_IS_EDITOR_VERSION)) {
+      return frozen.getProperty(WOPIService.MSOFFICE_IS_EDITOR_VERSION).getBoolean();
+    }
+    return false;
+  }
 
+  /**
+   * Creates a version of draft. Used to create version after manually uploaded
+   * content.
+   *
+   * @param node the node
+   * @throws RepositoryException the repository exception
+   * @throws OfficeOnlineException the OfficeOnlineException exception
+   */
+  protected void createVersionOfDraft(Node node) throws RepositoryException, OfficeOnlineException {
+    ConversationState contextState = ConversationState.getCurrent();
+    SessionProvider contextProvider = sessionProviders.getSessionProvider(null);
+    String userId = node.getProperty(EXO_LAST_MODIFIER).getString();
+    if (setUserConvoState(userId)) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Creating a version from draft. Path: " + node.getPath() + " user: " + userId);
+      }
+      try {
+        node.save();
+        if (checkout(node)) {
+          node.checkin();
+          node.checkout();
+        }
+      } catch (Exception e) {
+        LOG.error("Couldnl't create a version from draft for user: " + userId);
+      }
+    } else {
+      throw new OfficeOnlineException("Cannot set conversation state " + userId);
+    }
+    restoreConvoState(contextState, contextProvider);
```

Comment:
Should it be in finally block to be sure we restore in any case?

Reply:
We need to restore a conversation state in the case when it was set up. 
setUserConvoState(..) doesn't throw any exceptions.
If we failed to set a convo state - we throw the OfficeOnlineException, else - restore the state. So no need to put it into the finally block.